swig/include/python/gdal_python.i

```diff
@@ -5165,6 +5165,117 @@ def quiet_warnings():
+
+       .. versionadded: 3.11
+
+       Parameters
+       ----------
+       algorithm: str or list[str]
+            Path to the algorithm. For example "raster info", or ["raster", "info"].
+       arguments: dict
+            Input arguments of the algorithm. For example {"format": "json", "input": "byte.tif"}
+       progress: callable
+            Progress function whose arguments are a progress ratio, a string and a user data
+       optimize_single_output: bool
+            Whether to return a single value when there is a single output argument, and to
+            deserialize automatically JSON responses as a dict, and return a osgeo.gdal.Dataset
+            when possible.
+       finalize: bool
+            Whether to run :py:func:`gdal.Algorithm.Finalize` when the context is released.
+       kwargs:
+            Instead of using the ``arguments`` parameter, it is possible to pass
+            algorithm arguments directly as named parameters of gdal.run().
+            If the named argument has dash characters in it, the corresponding
+            parameter must replace them with an underscore character.
+            For example ``dst_crs`` as a a parameter of gdal.run(), instead of
+            ``dst-crs`` which is the name to use on the command line.
+
+       Returns
+       -------
+            A context manager with the output arguments of the algorithm
+
+       Example
+       -------
+
+       >>> with gdal.run(["raster", "info"], {"input": "byte.tif"}) as res:
+       ...     print(res["bands"])
+
+       >>> with gdal.run("raster reproject", input="byte.tif", output_format="MEM", dst_crs="EPSG:4326") as ds
+       ...     print(ds.ReadAsArray())
+    """
+
+    if isinstance(algorithm, list):
+        alg = GetGlobalAlgorithmRegistry()
+        parent_name = "gdal"
+        for i, v in enumerate(algorithm):
+            if i == 0 and v == "gdal":
+                continue
+            subalg = alg[v]
+            if not subalg:
+                raise Exception(f"{v} is not a valid sub-algorithm of {parent_name}")
+            alg = subalg
+            parent_name = alg.GetName()
+    elif isinstance(algorithm, str):
+        if algorithm.startswith("gdal "):
+            algorithm = algorithm[len("gdal "):]
+        alg = GetGlobalAlgorithmRegistry()
+        parent_name = "gdal"
+        for i, v in enumerate(algorithm.split(' ')):
+            if i == 0 and v == "gdal":
+                continue
+            subalg = alg[v]
+            if not subalg:
+                raise Exception(f"{v} is not a valid sub-algorithm of {parent_name}")
+            alg = subalg
+            parent_name = alg.GetName()
+    else:
+        raise Exception("Wrong type for algorithm. Expected string or list of string")
+
+    for k in arguments:
+        alg[k.replace('_', '-')] = arguments[k]
+
+    for k in kwargs:
+        alg[k.replace('_', '-')] = kwargs[k]
+
+    assert alg.Run(progress)
```

Comment:
Probably want to throw a `RuntimeError` here and when checking `alg.Finalize()` below.